src/mlpack/methods/decision_tree/decision_tree.hpp

```diff
@@ -0,0 +1,235 @@
+/**
+ * @file decision_tree.hpp
+ * @author Ryan Curtin
+ *
+ * A generic decision tree learner.  Its behavior can be controlled via template
+ * arguments.
+ */
+#ifndef MLPACK_METHODS_DECISION_TREE_DECISION_TREE_HPP
+#define MLPACK_METHODS_DECISION_TREE_DECISION_TREE_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "gini_gain.hpp"
+#include "best_binary_numeric_split.hpp"
+#include "all_categorical_split.hpp"
+
+namespace mlpack {
+namespace tree {
+
+/**
+ * This class implements a generic decision tree learner.  Its behavior can be
+ * controlled via its template arguments.
+ *
+ * The class inherits from the auxiliary split information in order to prevent
+ * an empty auxiliary split information struct from taking any extra size.
+ */
+template<typename FitnessFunction = GiniGain,
+         template<typename> class NumericSplitType = BestBinaryNumericSplit,
+         template<typename> class CategoricalSplitType = AllCategoricalSplit,
+         typename ElemType = double,
+         bool NoRecursion = false>
+class DecisionTree :
+    public NumericSplitType<FitnessFunction>::template
+        AuxiliarySplitInfo<ElemType>,
+    public CategoricalSplitType<FitnessFunction>::template
+        AuxiliarySplitInfo<ElemType>
+{
+ public:
+  //! Allow access to the numeric split type.
+  typedef NumericSplitType<FitnessFunction> NumericSplit;
+  //! Allow access to the categorical split type.
+  typedef CategoricalSplitType<FitnessFunction> CategoricalSplit;
+
+  /**
+   * Construct the decision tree on the given data and labels.
+   */
+  template<typename MatType>
+  DecisionTree(const MatType& data,
```

Comment:
Ah, yes, let me go through and update all of the documentation to make sure it is comprehensive.